ptbcontrib/extract_passed_user/extract.py

```diff
@@ -0,0 +1,113 @@
+#!/usr/bin/env python
+#
+# A library containing community-based extension for the python-telegram-bot library
+# Copyright (C) 2020-2025
+# The ptbcontrib developers
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Lesser Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser Public License for more details.
+#
+# You should have received a copy of the GNU Lesser Public License
+# along with this program.  If not, see [http://www.gnu.org/licenses/].
+
+"""This module attempts to extract side user passed actual user from incoming message"""
+
+from __future__ import annotations
+
+import asyncio
+from typing import TYPE_CHECKING, Callable
+
+from telegram import SharedUser
+from telegram.error import TelegramError
+
+from ..username_to_chat_api import UsernameToChatAPI
+
+if TYPE_CHECKING:
+    from telegram import Chat, Message
+
+
+def get_nums_from_text(
+    text: str,
+) -> int | None:
+    """Extract numbers from text"""
+    try:
+        return int("".join([letter for letter in text if letter.isdigit()]))
+    except ValueError:
+        return None
+
+
+def default_resolver(
+    wrapper: UsernameToChatAPI,
+    username: str,
+) -> Chat | None:
+    """
+    Try to resolve username to Chat object.
+    Catch exceptions from username_to_chat_api ?
+    """
+    return asyncio.run(
+        wrapper.resolve(
+            username=username.strip(),
+        )
+    )
+
+
+async def extract_passed_user(
+    message: Message,
+    username_resolver: Callable | UsernameToChatAPI | None = None,
+) -> SharedUser | None:
+    """
+    4 cases:
+    1. message.users_shared
+    2. message.contact.user_id
+    3. message.text starts with '@' and resolved by wrapper
+    4. message.text has only numbers, resolved by get_nums_from_text
+
+    About contact entity:
+        1. `contact` may be without user_id.
+        2. `contact` not contain @name at all, so not convertable to complete SharedUser.
+    """
+    chat = user_id = None
+    if message.users_shared:  # Note: May be select multiple
+        return message.users_shared.users[0]
+    if username_resolver and message.text and message.text.strip().startswith("@"):
+        if isinstance(username_resolver, UsernameToChatAPI):
+            chat = default_resolver(
+                wrapper=username_resolver,
+                username=message.text,
+            )
+        else:
+            chat: Chat | None = await username_resolver(  # type: ignore[no-redef]
+                username=message.text,
+            )
+    elif hasattr(
+        message.contact,
+        "user_id",
+    ):
```

Comment:
this will always be `True`. I guess you'd rather want to check `if message.cont
```suggestion
    elif message.contact:
```

Reply:
Are you sure? In most cases, `message.contact` is `None`. However, even if it isn't, the `contact` object might not have a `user_id` field, but it will still be a valid contact. So the expression `hasattr(message.contact, "user_id")` checks both these things at the same time.

Reply:
In most cases message.contact is None, however, even if it isn't, the contact object might not have a user_id field, and it will still be a valid contact, but I really missed the case with `user_id` intentionally filled with `None`.

Reply:
`contact.user_id` will always be there - it just may be `None`. So `if message.contact and message.contact.user_id` might be what you need.